pkg/controller/pxc/controller.go

```diff
@@ -534,6 +476,111 @@ func (r *ReconcilePerconaXtraDBCluster) Reconcile(request reconcile.Request) (re
 	return rr, nil
 }
 
+func (r *ReconcilePerconaXtraDBCluster) reconcileHAProxy(cr *api.PerconaXtraDBCluster, owner metav1.OwnerReference) error {
+	if !cr.HAProxyEnabled() {
+		if err := r.disableHAProxyReplicasService(cr); err != nil {
+			return errors.Wrap(err, "delete HAProxy replica service")
+		}
+
+		if err := r.disableHAProxyService(cr); err != nil {
+			return errors.Wrap(err, "delete HAProxy service")
+		}
+
+		if err := r.disableHAProxy(cr); err != nil {
+			return errors.Wrap(err, "delete HAProxy stateful set")
+		}
+
+		return nil
+	}
+
+	if err := r.enableHAProxy(cr); err != nil {
+		return errors.Wrap(err, "HAProxy upgrade error")
+	}
+
+	if err := r.enableHAProxyService(cr, owner); err != nil {
+		return errors.Wrap(err, "failed to create or update haproxy service")
+	}
+
+	if !cr.HAProxyReplicasServiceEnabled() {
+		if err := r.disableHAProxyReplicasService(cr); err != nil {
+			return errors.Wrap(err, "delete HAProxy replica service")
+		}
+
+		return nil
+	}
+
+	if err := r.enableHAProxyReplicasService(cr, owner); err != nil {
+		return errors.Wrap(err, "failed to create or update haproxy replicas")
+	}
+
+	return nil
+}
+
+func (r *ReconcilePerconaXtraDBCluster) enableHAProxy(cr *api.PerconaXtraDBCluster) error {
+	return r.updatePod(statefulset.NewHAProxy(cr), &cr.Spec.HAProxy.PodSpec, cr, nil)
+}
+
+func (r *ReconcilePerconaXtraDBCluster) enableHAProxyService(cr *api.PerconaXtraDBCluster, owner metav1.OwnerReference) error {
+	haProxyService := pxc.NewServiceHAProxy(cr, owner)
+	haProxyService.Spec.Type = corev1.ServiceTypeClusterIP
+
+	if len(cr.Spec.HAProxy.ServiceType) > 0 {
+		haProxyService.Spec.Type = cr.Spec.HAProxy.ServiceType
+	}
+
+	if haProxyService.Spec.Type == corev1.ServiceTypeLoadBalancer ||
```

Comment:
we have all this checks and values setup in `pxc.NewServiceHAProxy(cr,owner)`

Reply:
This code is added by you in https://github.com/percona/percona-xtradb-cluster-operator/pull/770/files#diff-3fa3ab828e096d1652e014a7a6325074f9cfbecf9894d9ca7e609600cfc9ccd0L383

I just moved it into separate function here. I'm confused what I should do.
However, I think it should be here because there is reason why it's here. And even when we just support single type today. We could extend it later. In this case, it could break logic.

Reply:
if you started to refactor the code, make sure to make a refactoring all parts, not only few of them.
Just check what checks already inside the creation function and delete them from there